##### [Forrige \>](1%20-%20Introduksjon.ipynb)  |     [Neste \>](3%20-%20Funksjoner.ipynb)

<span style="color:#9bb1d4;font-size:30px;font-weight:900"> 2 - Python basics </span>  

## Programmeringsspråk
Python, som de fleste andre programmeringsspråk, er basert på engelsk språk. Det er naturlig, siden engelsk er det overlegent største andrespråket i verden og de fleste programmeringsspråk har blitt utviklet i USA (her er Python et unntak, da nederlenderen Guido van Rossum er bak Python). 

Uansett, engelsk er dominerende og av samme grunn bør all kode i utgangspunktet skrives på engelsk. Kode bør kunne deles globalt, og da bør den skrives på et språk flest mulig forstår. Får man denne vanen inn tidlig, vil den følge din ferd som programmerer.

Men selv om koden er engelsk, så er ofte brukergrensesnittet på lokalt språk. Du har kanskje norsk som språk i nettleseren din, men bortsett fra navnene i menyene er all kode på engelsk. 

I dette kurset skal vi gjøre det på tilsvarende måte. All kode skal i hovedsak skrives på engelsk. Printing (og kommentarer) kan imidlertid gjøres på norsk, slik som dette:

In [2]:
# La oss lage et enkelt program
my_sentence = "Her er det lov å skrive norsk, for dette skal ut til skjermen!"
print(my_sentence)

# La oss skrive en ny setning
sentence = "Jeg heter Ola Nordmann"
print(sentence)

Her er det lov å skrive norsk, for dette skal ut til skjermen!
Jeg heter Ola Nordmann


Legg merke til den første setningen i programmet, `# La oss lage et enkelt program`. Hashtaggen (*emneknagg*) `#` først, angir at denne linjen ikke skal oppfattes som en del av programmet. Python overser dermed denne setningen. Denne funksjonen finnes, slik at vi kan *dokumentere* koden. Det vil si at vi skriver inn i koden hva som skjer. 

Program blir ofte veldig fort veldig komplekse og uoversiktlige. Uten noen forklaring blir det fort vanskelig eller uoverkommelig for andre å forstå hva du har tenkt. Da blir det veldig vanskelig for andre å komme inn eller ta over prosjektet ditt. Det er også enklere for deg selv å gå tilbake og lese din egen kode, om du har vært flink å dokumentere den med kommentarer. 

## Gode programmeringsvaner

For å gjøre koden så godt leselig som mulig for andre, er det noen prinsipper som bør følges. Det er lite hensiktsmessig å gå detaljert inn på disse nå, før vi har lært litt mer, men vi tar med de generelle prinsippene som skaperen av Python, Guido van Rossum har laget, som du kan kikke tilbake på etterhvert som du lærer mer:

1. Pen er bedre enn stygt.
2. Eksplisitt er bedre enn implisitt.
3. Enkelt er bedre enn komplekst.
4. Komplekst er bedre enn komplisert.
5. Flat er bedre enn nøstet.
6. Luftig er bedre enn tett.
7. Leselighet betyr noe.
8. Spesialtifeller er ikke spesielle nok til å bryte regelene.
   Selv om praktiske hensyn slår renhet.
9. Feil bør aldri passere i det stille,
   med mindre de er eksplisitt dempet.
10. Når du møter tvilstilfeller, ikke fall for fristelsen å gjette.
11. Det bør være én og helst bare én åpenbar måte å gjøre det på,
    selv om den måten kanskje ikke er åpenbar umiddelbart for de som ikke er Nederelendere
12. Nå er bedre aldri,
    selv om aldri ofte er bedre enn *akkurat* nå.
13. Om implementeringen er vanskelig å forstå, så er det en dårlig idé. 
14. Om implementeringen er lett å forstå, så er det en god idé.  
15. `namespaces` er en fantastisk idé, -- la oss gjøre mer av det!

## Variabler
Variablene i Python kan være ulike objekter som i alle andre programmeringsspråk, og typen objekt defineres av hva du setter variabelen til. Det er vanlig å definere stort sett alle variabler med små bokstaver i Python. Kun unntaksvis brukes store bokstaver.

Du kan ikke bruke mellomrom i variabelnavn. Det vil oppfattes som flere ulike navn. For eksempel er `number_of_apples`  tillatt men ikke "number of apples". Vi bruker også kun engelske bokstaver, selv om norske fungerer. `aintail_æpla` kan altså brukes, men anbefales ikke.

In [3]:
# Her er det flere, ulike variabler

# Heltall
variable_a = 1

# Desimaltall
variable_b = 2.5678

# Sant/usant
variable_c = True

# Tekst
variable_d = "Hello, World!"

## ```bool```(Sann/usann)

Den enkleste variabelen i Python er den binære typen "bool". Den kan ta to verdier, `True` eller `False` (sant eller usant). Denne variabelen brukes mest i sammenhenger der vi skal teste om et vilkår er oppfylt eller ikke. Da bruker vi `if`-setninger, som vi skal komme tilbake til.

Vi kan sjekke hvilken type en variabel er med funksjonen type(). Under får vi bekreftet `apple_is_tasty` er en boolsk variabel.

#### Eksempel 3:

In [4]:
# Eksempel på bool-variabel
apple_is_tasty = True
type(apple_is_tasty)

bool

In [5]:
# Printer verdien til 'apple_is_tasty'
print(apple_is_tasty)

True


Legg merke til hva som skjer her: Første linje er bare en kommentar, som nevnt tidligere. I andre linjer *setter* vi variabelen `apple_is_tasty` lik verdien `True`. Det er på denne måten vi definerer variabler i Python. Når vi har satt en variable på denne måten, bestemmer det hvilken `type` variabel det er. 

Det er ikke alltid at variabelen settes rett før vi skal bruke den. Er det et stort program kan variabelen ha vært satt på en helt annen plass i programmet. I så fall er det nyttig å vise hvilken type variabel `apple_is_tasty` representerer. Siden den er satt til verdien `True` (sant), er det en `bool`. 

Når vi kjører et kodefelt i Jupyter, får vi resultatet av den siste setningen skrevet ut til skjermen, etter kodefeltet. Om du har kjørt koden, vil du se at det står `bool`der. Det er fordi funksjonen `type()` står i siste linje, og den returnerer hvilken type `apple_is_tasty` er. 

I neste kodeblokk kjører vi funksjonen ```print()``` som rett og slett bare skriver ut hvilken *verdi* variabelen har (dette tilfelle ```True```).

## `int` (heltall)

```int``` betyr *integer* (heltall på norsk), og naturlig nok, representerer hele tall i Python. Under ser vi at Python tolker `8` som et heltall:

#### Eksempel 4:

In [6]:
# Heltall/integer eksempel

number_of_apples = 8
type(number_of_apples)

int

Vi kan konvertere `int` til `bool` og `bool` til `int` ved å bruke disse som funksjoner slik som dette:

#### Eksempel 5:

In [7]:
# Gjøre om variabletype
type(int(apple_is_tasty))

int

In [8]:
number = 8.00
print(f'{number} er av typen {type(number)}')

# Nå gjør vi om 'number'
number_two = int(number)
print(f'{number_two} er av typen {type(number_two)}')

8.0 er av typen <class 'float'>
8 er av typen <class 'int'>


Over ser vi at vi startet med tallet `8.00`, som vi ser er av typen `float`, så brukte vi funksjonen `int()` til å gjøre om `number` (som var lik `8.00`) til heltallet `8`.

In [9]:
type(bool(0))

bool

Legger merke til at i python så tilsvarer 1 `True` og 0 tilsvarer `False`, slik at

#### Eksempel 6:

In [10]:
20*False

0

In [11]:
20*True

20

## `float` (desimaltall)

Regning gjøres vanligvis med desimaltall, eller "floating point" som det kalles av på fagspråket.  I python heter derfor denne variabeltypen  `float`.

Om for eksempel 11 skal tolkes som et heltall, gir vi beskjed til Python om at det er et desimaltall ved å skrive det med én desimal. Alle programmeringsspråk, inkludert Python, bruker punktum som desimaltegn. 

#### Eksempel 7:

In [12]:
# Floating point (float), altså desimaltall

apple_price = 11.0
type(apple_price)

float

In [13]:
apple_price = 11
type(apple_price)

int

Legg merke til de to ulike kodeblokkene over - **hva er forskjellen her?**

Navnet "floating point", eller **flyttall** på norsk, kommer av at alle flyttall har eksakt like mange siffer, men desimaltegnet kan flyttes slik at `float` kan representere et svært stort spenn av verdier. Det største flyttallet er én med 308 nuller etter. Desimaltegnet flyttes enkelt fordi at `float` egentlig er på "vitenskapelig form" (se under), selv om det ofte vises som vanlig tall. For eksempel så kan vi skrive tallet 102 på vitenskapelig form, men Python spytter det ut på et format vi enklere kan forstå

#### Eksempel 8:

In [14]:
insane_apple_price = 1.02e+2
print(insane_apple_price)

102.0


Vitenskapelig form `1.02e+2` er en kombinasjon av et heltall (`1`), desimaler (`.02`) og hvilken potens av 10 tallet skal multipliseres med (`e+2`).  

`1.02e+2` betyr altså at $1.02$ skal multipliseres med  $10^{2}=100$. 

`1.02e+2` betyr dermed $1.02\cdot 10^{2}=102$. På samme kan $0.0102$ skrives `1.02e-2`,  eller  $1.02\cdot 10^{-2}=102$.

Legg også merke til at funksjonen `print()`er brukt her. Det er ofte bedre å printe eksplisitt med printfunksjonen. Dersom det du ønsker å se ikke står sist i koden er det også nødvendig å bruke `print()`. 

Her er et eksemple med et lite flyttall:

#### Eksempel 9:

In [15]:
really_cheap_apple_price = 1.02e-2
print(really_cheap_apple_price)

0.0102


Når vi skriver ut resultatet får vi det altså ut som `102.0` eller `0.0102` i stedet for det vitenskapelige flyttallformatet. Det er fordi Python skjønner at `1.02e+2` og  `1.02e-2` ikke ser så pent ut for oss. Python formaterer derfor tallet til et mer lesbart format før det skrives ut.

Det er enkelt å regne med både `float` og  `int`:

#### Eksempel 10:

In [16]:
total = apple_price*number_of_apples
print(total)

88


Heldigvis spiller det liten rolle om noe er `int` eller `float` når du skal regne. I de fleste andre programmeringsspråk får du et nytt heltall når du deler et heltall på et annet. I Python gir dette et desimaltall:


#### Eksempel 11:

In [17]:
number_of_persons = 5
apple_per_person  = number_of_apples/number_of_persons
print(apple_per_person)

1.6


In [18]:
5/5

1.0

## Regning med `float`

For de fire regneartene fungerer `float` slik vi har lært på barneskolen:

#### Eksempel 12:

In [19]:
print(2+3)
print(2-3)
print(2*3)
print(2/3)

5
-1
6
0.6666666666666666


Ofte har vi imidlertid bruk for å opphøye et tall i et annet. Om vi for eksempel vil vite hva $2\cdot2\cdot2$ er, så kan vi istedet skrive $2^3$. Dette kalles en *potens* med 2 som *grunntall* og 3 som *eksponent*. I python utføres denne operasjonen med dobbelt gange-tegn `**`. Dette er ulikt endel andre programmeringsspråk, der *cirkumfleks*-symbolet `^` ofte brukes. 

Slik regner du med potens i Python:

#### Eksempel 13:

In [20]:
2**3

8

### `str` (tekst)

Tekst kalles "streng" på dataspråket, eller "string" på engelsk. Typen som representerer tekst heter derfor `str`. Vi forteller Python at dette er en tekst ved å sette anførselstegn rundt. Her er et eksempel på en streng:

#### Eksempel 14:

In [21]:
# string:
x = "Dette er en streng (dvs tekst)"
type(x)

str

Det er enkelt å sette sammen `string` og andre variabler i Python. Det er flere måter å gjøre det på, men vi skal bruke en metode som er ganske ny i Python. Fra og med versjon 3.6 kom noe som kalles "f-strings" som gir oss en veldig enkel måte å blandet tekst og andre variabler. Her er et eksempel:

#### Eksempel 15:

In [22]:
print(f"Det er {apple_is_tasty} at jeg synes epler smaker godt.")

Det er True at jeg synes epler smaker godt.


Legg merke til f'en før det første anførselstegnet. Den angir at dette er en f-streng og at alt som er inne i klammeparenteser {} skal omformateres til `string` og tas med i teksten. 

Et alternativ er å bruke `+`-tegnet. Det går nemlig an å summere tekst i Python:

#### Eksempel 16:

In [23]:
"Jeg "+"gikk "+"derfor "+"på "+"butikken "+"for "+"å "+"kjøpe"

'Jeg gikk derfor på butikken for å kjøpe'

Men det går ikke an å summere tekst og andre variabler, da skjønner ikke Python hva du ønsker å oppnå

#### Eksempel 17:

In [24]:
"Jeg gikk derfor på butikken for å kjøpe " + number_of_apples + " epler. "

TypeError: can only concatenate str (not "int") to str

Du må i så fall omformatere tallet `number_of_apples` til streng:

#### Eksempel 18:

In [25]:
"Jeg gikk derfor på butikken for å kjøpe " + str(number_of_apples) + " epler."

'Jeg gikk derfor på butikken for å kjøpe 8 epler.'

Derfor er det langt enklere å bruke f-strings. Dersom tekststrengen er for lang for én linje, kan du sette flere tekstrenger etter hverandre på hver sin linje, slik som dette:

#### Eksempel 19:

In [26]:
print(
    f"Det er {apple_is_tasty} at jeg synes epler smaker godt. \n"
    f"Jeg gikk derfor på butikken for å kjøpe {number_of_apples} epler. \n"
    f"Innehaveren gav meg en vanvittig pris. Hun skulle ha {insane_apple_price} kroner per eple. \n"
    f"Jeg fortalte henne at det var helt uaktuelt, og foreslo {really_cheap_apple_price} kroner. \n"
    f"Hun avslo, men vi ble til sist enig om {apple_price} kroner.\n"
    f"Jeg betalte hennet til sammen {apple_price*number_of_apples} kroner og dro hjem. \n\n"
    f"På veien traff jeg {number_of_persons} venner. Jeg hadde ikke noe annet valg enn å dele eplene mine med dem. \n"
    f"Det ble {number_of_apples/number_of_persons} epler per person.\n"
)

Det er True at jeg synes epler smaker godt. 
Jeg gikk derfor på butikken for å kjøpe 8 epler. 
Innehaveren gav meg en vanvittig pris. Hun skulle ha 102.0 kroner per eple. 
Jeg fortalte henne at det var helt uaktuelt, og foreslo 0.0102 kroner. 
Hun avslo, men vi ble til sist enig om 11 kroner.
Jeg betalte hennet til sammen 88 kroner og dro hjem. 

På veien traff jeg 5 venner. Jeg hadde ikke noe annet valg enn å dele eplene mine med dem. 
Det ble 1.6 epler per person.



Merk at setningene må stå inne i en parentes, for eksempel som argument i en funksjon som over, for at Python skal forstå at de skal settes sammen til én setning. Legg også merke til `\n`på slutten av hver delsetning. Denne indikerer at det skal være linjeskift etter hver delsetning. 

## Anførselstegn for strenger

Det finnes tre forskjellige typer tegn for å markere at noe er tekst i Python. Disse er apostrof `'`, anførselstegn `"` og trippel anførselstegn `"""`. Rollen til de to første typene er å enkelt kunne bruke anførselstegn inne i teksten:

#### Eksempel 20:

In [27]:
print( "Jeg gikk derfor på butikken for å 'kjøpe' epler.")
print( 'Jeg gikk derfor på butikken for å "kjøpe" epler.')

Jeg gikk derfor på butikken for å 'kjøpe' epler.
Jeg gikk derfor på butikken for å "kjøpe" epler.


Trippel anførselstegn `"""` brukes når du vil ha tekst over flere linjer, for eksempel

#### Eksempel 21:

In [28]:
s=f"""Det er {apple_is_tasty} at jeg synes epler smaker godt.
Jeg gikk derfor på butikken for å kjøpe {number_of_apples} epler.
Innehaveren gav meg en vanvittig pris. Hun skulle ha {insane_apple_price} kroner per eple.
Jeg fortalte henne at det var helt uaktuelt, og foreslo {really_cheap_apple_price} kroner.
Hun avslo, men vi ble til sist enige om {apple_price} kroner.
Jeg betalte hennet til sammen {apple_price*number_of_apples} kroner og dro hjem.

På veien traff jeg {number_of_persons} venner. Jeg hadde ikke noe annet valg enn å dele eplene mine med dem.
Det ble {number_of_apples/number_of_persons} epler per person."""
print(s)

Det er True at jeg synes epler smaker godt.
Jeg gikk derfor på butikken for å kjøpe 8 epler.
Innehaveren gav meg en vanvittig pris. Hun skulle ha 102.0 kroner per eple.
Jeg fortalte henne at det var helt uaktuelt, og foreslo 0.0102 kroner.
Hun avslo, men vi ble til sist enige om 11 kroner.
Jeg betalte hennet til sammen 88 kroner og dro hjem.

På veien traff jeg 5 venner. Jeg hadde ikke noe annet valg enn å dele eplene mine med dem.
Det ble 1.6 epler per person.


Alternativt kan du sette hver setning på linje etterhverandre med egne enkel anførselstegn, som i Eksempel 219, men da må alle setningene befinne seg inne i en parentes.

# Oppgaver
Nedenfor her vil dere finne flere oppgaver som gjør seg godt for å bli kjent med de ulike variabelene vi har nevnt.

#### Oppgave 1

Bruk koden i Eksempel 19, men fjern `\n` og se hva som skjer

In [29]:
print(
    f"Det er {apple_is_tasty} at jeg synes epler smaker godt. "
    f"Jeg gikk derfor på butikken for å kjøpe {number_of_apples} epler. "
    f"Innehaveren gav meg en vanvittig pris. Hun skulle ha {insane_apple_price} kroner per eple. "
    f"Jeg fortalte henne at det var helt uaktuelt, og foreslo {really_cheap_apple_price} kroner. "
    f"Hun avslo, men vi ble til sist enig om {apple_price} kroner. "
    f"Jeg betalte hennet til sammen {apple_price*number_of_apples} kroner og dro hjem. "
    f"På veien traff jeg {number_of_persons} venner. Jeg hadde ikke noe annet valg enn å dele eplene mine med dem. "
    f"Det ble {number_of_apples/number_of_persons} epler per person."
)

Det er True at jeg synes epler smaker godt. Jeg gikk derfor på butikken for å kjøpe 8 epler. Innehaveren gav meg en vanvittig pris. Hun skulle ha 102.0 kroner per eple. Jeg fortalte henne at det var helt uaktuelt, og foreslo 0.0102 kroner. Hun avslo, men vi ble til sist enig om 11 kroner. Jeg betalte hennet til sammen 88 kroner og dro hjem. På veien traff jeg 5 venner. Jeg hadde ikke noe annet valg enn å dele eplene mine med dem. Det ble 1.6 epler per person.


### 🔍 Hva skjer når vi fjerner `\n`?
* `\n` betyr linjeskift, så når vi fjerner det, blir all teksten skrevet ut på én lang linje i stedet for å være pent formatert på flere linjer.
* Det blir altså mye vanskeligere å lese outputtet.

#### Oppgave 2
Lag variabler som du putter inn i en historie som i Eksempel 19

In [31]:
## Her kan man gjøre akkurat det samme men fjerne variablene, og sette inn dine nye.

# Definerer variabler
navn = "Ola"
sted = "parken"
antall_venner = 3
aktivitet = "spilte fotball"
værtype = "solfylt"
snack = "is"
pris_per_is = 25
totalpris = antall_venner * pris_per_is

# Skriver ut historien med variablene, hvor vi bruker \n for linjeskift
print(
    f"En dag dro {navn} til {sted} fordi det var en {værtype} dag.\n"
    f"Sammen med {antall_venner} venner {aktivitet} de i flere timer.\n"
    f"Etterpå ble de sultne og bestemte seg for å kjøpe {snack}.\n"
    f"Hver {snack} kostet {pris_per_is} kroner, så til sammen betalte de {totalpris} kroner.\n"
    f"Det var en flott dag i {sted}!"
)

En dag dro Ola til parken fordi det var en solfylt dag.
Sammen med 3 venner spilte fotball de i flere timer.
Etterpå ble de sultne og bestemte seg for å kjøpe is.
Hver is kostet 25 kroner, så til sammen betalte de 75 kroner.
Det var en flott dag i parken!


### 🔍 Hva har vi gjort?
* Variablene lagrer informasjon som kan endres til andre verdier senere.
* `f""` brukes for å sette variabler inn i en streng.
* Koden lager en enkel historie som kan tilpasses ved å endre variablene.

🚀 Nå kan du endre slik du måtte ønske selv!

#### Oppgave 3
a) Hvilke variabler kan brukes sammen med multiplikasjonsoperatoren `*`? 
prøv for eksempel `"hei"*2.0` og `True*2.0`

In [32]:
# Gyldige kombinasjoner

# Tall
print(5 * 2)      # 10  (int * int)
print(5.0 * 2)    # 10.0  (float * int)
print(5 * 2.0)    # 10.0  (int * float)

# Streng
print("hei" * 2)  # "heihei"
print("abc" * 3)  # "abcabcabc"

# Boolean
print(True * 2)   # 2  (True tolkes som 1)
print(False * 5)  # 0  (False tolkes som 0)

10
10.0
10.0
heihei
abcabcabc
2
0


In [33]:
# Ugyldige kombinasjoner
print("hei" * 2.0)  

TypeError: can't multiply sequence by non-int of type 'float'

**Obs:** her feilet noe. Men les feilmeldingen, altså det som står etter `TypeError` - hva sier denne oss?

b) Hvilke variabler kan brukes sammen med addisjonsoperatoren `+`?

In [ ]:
# Gyldige operasjoner med tall
print(5 + 3)      # 8  (int + int)
print(4.5 + 2.5)  # 7.0  (float + float)
print(10 + 2.3)   # 12.3  (int + float)

# Gyldige operasjoner med tekst
print("Hei, " + "verden!")  # "Hei, verden!"
navn = "Ola"
print("Hei, " + navn)  # "Hei, Ola"

# Kan også så slammen lister (men det skal vi ser mer på senere)
print([1, 2, 3] + [4, 5, 6])  # [1, 2, 3, 4, 5, 6]

In [34]:
# Ugyldig bruk av +
print("Alder: " + 25)

TypeError: can only concatenate str (not "int") to str

Løsningen for å kjøre eksempelet over er å gjøre om `int` til `str`:

```python
print("Alder: " + str(25))  # "Alder: 25"
```
**Prøv dette selv.**

🚀 Konklusjon: `+` fungerer med tall, strenger, lister og tupler, men ikke på tvers av inkompatible typer.

c) Hvilke variabler kan brukes sammen med divisjonsoperatoren `/`?

In [35]:
# Gyldige operasjoner
print(10 / 2)  # 5.0 (int / int gir float)
print(9.0 / 2.0)  # 4.5
print(10 / 3)     # 3.3333333333333335
print(10 / True)   # 10.0 (10 / 1)

5.0
4.5
3.3333333333333335
10.0


In [36]:
# Ugyldige operasjoner
print(5.0 / False) # ❌ Feil: Division by zero!

ZeroDivisionError: float division by zero

In [37]:
print("Hei" / 2)

TypeError: unsupported operand type(s) for /: 'str' and 'int'

In [38]:
print(10 / 0)  # ❌ Feil

ZeroDivisionError: division by zero

d) Hvilke variabler kan brukes sammen med subtraksjonsoperatoren `-`?

In [40]:
# Gyldige operasjoner

# Vanlig subtraksjon
print(10 - 5)  # 5
print(7.5 - 2.2)  # 5.3

# Husk at True = 1 og False = 0
print(10 - True)   # 9 (10 - 1)
print(5.5 - False) # 5.5 (5.5 - 0)

# Kan gjøre om negative tall til positive, og omvendt
print(-10)   # -10
print(-(-5)) # 5

5
5.3
9
5.5
-10
5


In [39]:
# Ugyldig
print("Hei" - "H")  

TypeError: unsupported operand type(s) for -: 'str' and 'str'

#### Oppgave 4

a) Forsøk med prøving og feiling å finne det størst mulige `float`-tallet som ikke blir `inf`

In [41]:
# Dette er ved hjelp av en løkke, for å se største verdi ved å doble tallet 1, doble det igjen, og igjen og igjen så langt det går før det blir inf ("uendelig" for PC/MAC-en).

x = 1.0
while x * 2 != float("inf"):  # Så lenge det ikke blir "inf", fortsett å doble
    x *= 2

print(x)  # Siste gyldige verdi før "inf"

8.98846567431158e+307


In [42]:
# Kjører dere denne kodeblokken ser dere den høyeste verdien vi kan ha i python.
import sys
print(sys.float_info.max)  

1.7976931348623157e+308


🚀 Konklusjon:
Det største `float`-tallet i Python er omtrent $1.7976931348623157 × 10^{308}$.
Alt større enn dette blir `inf`.

b) Forsøk med prøving og feiling å finne `float`-tallet nærmest null som ikke blir `0.0`


In [43]:
x = 1.0  
while x / 2 > 0:  # Så lenge verdien ikke blir 0.0, fortsett å halvere
    x /= 2  

print(x)  # Siste gyldige verdi før den blir 0.0

5e-324


#### Oppgave 5

1. Hva er hensikten med en boolsk variabel?

Svar: En boolsk variabel brukes til å lagre en sannhetsverdi: enten `True` eller `False`. Det er nyttige i programmering for å si noe om man skal gjøre ditt eller datt, betingelser, løkker og andre logiske operasjoner. Dette vil vi bruke mye senere i kurset.

2. Hvordan kan du konvertere:
    * en boolsk verdi til en heltall verdi i Python? \n
    * et heltall til en streng i Python?
    * et flyttall til et heltall i Python?

In [46]:
# Konvertere boolsk verdi til heltall
bool_verdi = True
heltall = int(bool_verdi)  # Dette gir 1
heltall

1

In [47]:
heltall = 42
streng = str(heltall)  # Dette gir '42'
streng

'42'

In [48]:
flyttall = 3.14
heltall = int(flyttall)  # Dette gir 3

3. Hvorfor bruker vi f-strings i Python, og hvordan bruker du dem for å kombinere tekst og variabler?
f-strings er nyttige fordi det gjør det svært enkelt å kombinere ulike variabeltyper med tekst, på en pen og oversiktlig måte, uten å trenge å konvertere variablene først.

**Hvorfor bruke f-strings?** 

* **Lesbarhet**: F-strings gjør koden mer lesbar og enklere å forstå, fordi de lar deg inkludere variabler direkte i strengene.
* **Effektivitet**: F-strings er raskere enn andre metoder for strengformatering som `str.format()` og `%`-operatoren.
* **Enkelhet**: De tillater at du kan inkludere variabler og uttrykk direkte i strengen uten å måtte gjøre ekstra konverteringer eller bruke spesielle syntaksmetoder.

For å bruke en f-string, setter du et "f" foran en streng og bruker krøllparenteser `{}` for å plassere variablene eller uttrykkene du ønsker å inkludere i strengen.

In [49]:
# Eksempel
navn = "Alice"
alder = 25
melding = f"Hei, jeg heter {navn} og jeg er {alder} år gammel."
print(melding)
# Output: Hei, jeg heter Alice og jeg er 25 år gammel.

Hei, jeg heter Alice og jeg er 25 år gammel.


4. Hvordan representerer du desimaltall i Python, og hva er spesielt med hvordan de lagres i minnet?

Desimaltall i Python er flyttall, som er en tallrekke med 16 tall og et desimaltegn (".") etterfulgt av "1e-xxx"/"1e+xxx", som bestemmer hva tallet skal multipliseres med. xxx angir hvor mange ganger 10 skal opphøyes, så 1e-2 er for eksempel 0.01 og 2e+2 er 200.

I Python representeres desimaltall som **flyttall** (eller **floats**), som er tall med en desimaldel. Disse er i utgangspunktet lagret ved hjelp av **standard flyttallsformatet IEEE 754** som brukes til å representere desimaltall på en binær måte.

In [54]:
x = 3.14  # 3.14
y = 0.1   # 0.1

In [ ]:
a = 1e-2  # 1 * 10^-2, som er 0.01
b = 2e+2  # 2 * 10^2, som er 200

### **Spesielt med hvordan desimaltall lagres i minnet**
1. **Flyttallsrepresentasjon (IEEE 754)**:

* Flyttall i Python lagres i henhold til **IEEE 754-standard**. Dette betyr at de består av tre deler:
    * **Signifikant (mantissa)**: Den faktiske verdien som skal lagres.
    * **Eksponent**: Hvor mange ganger tallet skal multipliseres med 10.
    * **Fortegn**: Om tallet er positivt eller negativt.
Denne representasjonen tillater lagring av både veldig store og veldig små tall, men den har også noen **begrensninger**.

2. **Presisjonsproblemer**:

Flyttall kan ikke alltid representere desimaltall nøyaktig i binær form. Dette kan føre til små avrundingsfeil i beregningene. For eksempel er verdien `0.1` ikke representert eksakt i binær form, så når du jobber med flyttall kan du oppleve små unøyaktigheter.


In [55]:
print(0.1 + 0.2)

# Se på output av denne kodeblokken!

0.30000000000000004


In [60]:
print(round(0.1 + 0.2, 2)) # Runder av til 0.3

0.3


5. Hva betyr det å heve et tall til en potens, og hvordan utfører du denne operasjonen i Python?

Å heve et tall i en potens betyr å gange tallet med seg selv et angit antall ganger. Til dette brukes `**`. For eksempel er `10**2=10*10` og `100**0.5 = 10`.

In [61]:
resultat = 10 ** 2
print(resultat)  # Output: 100

100


In [64]:
resultat = 10 ** -2
print(resultat)

0.01


In [65]:
resultat = 10e+2 # Det samme som 10 x 10^2
print(resultat)

1000.0


#### Oppgave 6

1. Skriv en Python-kode som definerer variablene: name, age, og city med dine egne detaljer. Bruk deretter en f-string for å skrive en setning som kombinerer disse detaljene.

In [66]:
# Definerer variablene med egne detaljer
name = "John"
age = 30
city = "Oslo"

# Bruker en f-string for å kombinere detaljene
setning = f"Hei, jeg heter {name}, jeg er {age} år gammel, og jeg bor i {city}."

# Skriver ut setningen
print(setning)

Hei, jeg heter John, jeg er 30 år gammel, og jeg bor i Oslo.


2. Lag en kalkulator ved å definere to variabler `x` og `y`først. Koden skal viser summen, differansen, produktet, forholdet og `x` i potens av `y`.

In [67]:
# Definerer variablene
x = 10
y = 2

# Beregner operasjonene
sum_resultat = x + y
differanse_resultat = x - y
produkt_resultat = x * y
forhold_resultat = x / y
potens_resultat = x ** y

# Skriver ut resultatene
print(f"Summen av {x} og {y} er: {sum_resultat}")
print(f"Differansen mellom {x} og {y} er: {differanse_resultat}")
print(f"Produktet av {x} og {y} er: {produkt_resultat}")
print(f"Forholdet av {x} og {y} er: {forhold_resultat}")
print(f"{x} hevet til {y} er: {potens_resultat}")

Summen av 10 og 2 er: 12
Differansen mellom 10 og 2 er: 8
Produktet av 10 og 2 er: 20
Forholdet av 10 og 2 er: 5.0
10 hevet til 2 er: 100


3. Bruk boolske verdier for å avgjøre om et tall er oddetall eller partall, og skriv ut resultatet for tallet 15.

In [68]:
# Definerer tallet
tall = 15

# Sjekker om tallet er partall eller oddetall ved hjelp av modulo
# modulo (%) sjekker om det er delelig på to (altså partall) eller ikke (altså oddetall)
er_partall = (tall % 2 == 0)
er_oddetall = (tall % 2 != 0)

# Skriver ut resultatene
if er_partall:
    print(f"{tall} er et partall.")
elif er_oddetall:
    print(f"{tall} er et oddetall.")

15 er et oddetall.


5. Skriv en kode som definerer to variabler `var_type` og `x`. Du skal sette `var_type` til en datatype og `x` til en verdi, for eksempel slik som under. Koden skal konvertere `x` til datatypen angitt av `var_type`, og printe ut type og verdi før og etter konverering. Forsøkt så med ulike verdier for `x` og ulike datatyper i `var_type`.

In [70]:
# Definerer variablene
var_type = float  # Eksempel: ønsket datatype er float
x = "123.45"  # Eksempel: verdien er en streng

In [71]:
# Skriv ut type og verdi før konvertering
print(f"Før konvertering: type(x) = {type(x)}, verdi(x) = {x}")

# Konverterer x til den angitte datatypen
x = var_type(x)

# Skriv ut type og verdi etter konvertering
print(f"Etter konvertering: type(x) = {type(x)}, verdi(x) = {x}")

Før konvertering: type(x) = <class 'str'>, verdi(x) = 123.45
Etter konvertering: type(x) = <class 'float'>, verdi(x) = 123.45


## Hipphipp du er ferdig med disse oppgavene :D

##### [Forrige \>](1%20-%20Introduksjon.ipynb)  |     [Neste \>](3%20-%20Funksjoner.ipynb)